### PR TITLE
Add comments support

### PR DESCRIPTION
**Pull request by almasaeed2010**
- Creates a new node for each submission.
- Associates submissions with nodes by adding `nid` columns to the tripal_hq_submission table
- Creates a custom permissions access function to determine who can see the comments (admins can see all comments and users can see commented related their own submissions only).
- Adds a new column to the user/admin dashboards to view and add new comments.

### CODE REVIEW

**Comment by almasaeed2010** [comment]
We need to delete the node the submission is approved/rejected

**Comment by bradfordcondon** [comment]
>We need to delete the node the submission is approved/rejected

are you sure?

The node doesnt store the entity, right?  if so, whats the harm in leaving to preserve comments?  What about if someone proposes an edit later and that has comments? The comment history might be nice.

**Comment by almasaeed2010** [comment]
Yeah I thought of that. I am gonna change it to when the submission record it self is deleted then delete the comments.

File: includes/tripal_hq.api.inc [hunk 1]
```diff
@@ -239,3 +255,25 @@ function tripal_hq_get_curator_emails() {
   }
   return implode(', ', $mail);
 }
+
+/**
+ * Submission comments count.
+ *
+ * @param \stdClass $submission
```
[the remark this answers is not in the document]
Comment (almasaeed2010) [reply includes/tripal_hq.api.inc]:
```suggestion
 * @param \stdClass $submission
 *    The submission object.
```

File: includes/tripal_hq.api.inc [hunk 2]
```diff
@@ -239,3 +257,44 @@ function tripal_hq_get_curator_emails() {
...
+ *
+ * @return int
```
Comment (almasaeed2010) [inline includes/tripal_hq.api.inc]:
```suggestion
 * @return int
 *    The number of comments related to a given submission.
```

File: includes/tripal_hq.api.inc [hunk 3]
```diff
@@ -239,3 +257,44 @@ function tripal_hq_get_curator_emails() {
...
+  }
+
+  return 0;
+}
+
+/**
+ * Delete the comments node.
+ *
+ * @param \stdClass $submission
+ *     The submission object.
```
Comment (almasaeed2010) [inline includes/tripal_hq.api.inc]:
```suggestion
 *    The submission object.
```

File: includes/tripal_hq.api.inc [hunk 4]
```diff
@@ -239,3 +257,44 @@ function tripal_hq_get_curator_emails() {
...
+ *
+ * @return void
```
Comment (almasaeed2010) [inline includes/tripal_hq.api.inc]:
```suggestion

```

File: includes/tripal_hq.api.inc [hunk 5]
```diff
@@ -239,3 +257,44 @@ function tripal_hq_get_curator_emails() {
...
+ *    This function does not return a value.
```
Comment (almasaeed2010) [inline includes/tripal_hq.api.inc]:
```suggestion
 *   This function does not return a value.
```

File: includes/tripal_hq_admin_dashboard.form.inc [hunk 6]
```diff
@@ -46,6 +46,7 @@ function tripal_hq_admin_dashboard_form($form, &$form_state, $status = NULL) {
       'field' => 'created_at',
       'sort' => 'dsc',
     ],
+    'Comments',
```
Comment (almasaeed2010) [inline includes/tripal_hq_admin_dashboard.form.inc]:
```suggestion
    'Comments' => [
      'data' => t('Comments')
    ],
```

File: includes/tripal_hq_admin_dashboard.form.inc [hunk 7]
```diff
@@ -129,6 +131,9 @@ function tripal_hq_admin_dashboard_form($form, &$form_state, $status = NULL) {
       $bundle->label,
       $status,
       $created_at,
+      $submission->nid ? l(
+        'Add Comments (' . $comment_count . ')', 'node/' . $submission->nid
```
Comment (almasaeed2010) [inline includes/tripal_hq_admin_dashboard.form.inc]:
```suggestion
        l('Add Comments (' . $comment_count . ')'), 'node/' . $submission->nid
```

File: includes/tripal_hq.api.inc [hunk 8]
```diff
@@ -31,12 +33,10 @@ function tripal_hq_get_submission_by_id(int $submission_id) {
...
 
-  $results = db_select('public.tripal_hq_submission', 't')
-    ->fields('t')
-    ->condition('uid', $uid)
-    ->orderBy('id', 'desc')
-    ->execute()
-    ->fetchAll();
+  $results =
+    db_select('public.tripal_hq_submission', 't')->fields('t')->condition(
+      'uid', $uid
```
[the remark this answers is not in the document]
Comment (almasaeed2010) [reply includes/tripal_hq.api.inc]: I agree. I don't know why the drupal formatter is doing this.

File: includes/tripal_hq_admin_approve.form.inc [hunk 9]
```diff
@@ -1,12 +1,18 @@
 <?php
 
+/**
```
Comment (almasaeed2010) [inline includes/tripal_hq_admin_approve.form.inc]:
```suggestion
/**
/* Administrator form to approve requests.
/*
```

File: includes/tripal_hq_user_data.form.inc [hunk 10]
```diff
@@ -157,8 +164,7 @@ function tripal_hq_entity_form_validate($form, &$form_state) {
 
   try {
     $create_date = new DateTime($form_state['values']['author_date']);
-  }
-  catch (Exception $e) {
+  } catch (Exception $e) {
```
Comment (almasaeed2010) [inline includes/tripal_hq_user_data.form.inc]:
```suggestion
  } 
  catch (Exception $e) {
```

File: tripal_hq.module [hunk 11]
```diff
@@ -192,3 +197,99 @@ function tripal_hq_entity_delete($entity, $type) {
     tripal_hq_delete_submission($submission);
   }
 }
+
+/**
+ * @implements hook_node_info()
+ *
+ * @return array
```
Comment (almasaeed2010) [inline tripal_hq.module]:
```suggestion
 * @return array
 *   The list of nodes that this module creates.
```

File: tripal_hq.module [hunk 12]
```diff
@@ -192,3 +197,99 @@ function tripal_hq_entity_delete($entity, $type) {
...
+    ),
+    'has_title'   => FALSE,
+    'base'        => 'node_content',
+  ];
+
+  return $nodes;
+}
+
+/**
+ * @param string $title
```
Comment (almasaeed2010) [inline tripal_hq.module]:
```suggestion
 * Create a new node. 
 *
 * @param string $title 
 *    The title for the node.
```

File: tripal_hq.module [hunk 13]
```diff
@@ -192,3 +197,99 @@ function tripal_hq_entity_delete($entity, $type) {
...
+ *
+ * @return mixed
```
Comment (almasaeed2010) [inline tripal_hq.module]:
```suggestion
 * @return int
 *   The node id.
```

File: tripal_hq.module [hunk 14]
```diff
@@ -192,3 +197,99 @@ function tripal_hq_entity_delete($entity, $type) {
...
+
+/**
+ * Implements hook_node_access().
+ */
+function tripal_hq_node_access($node, $op, $account) {
+  global $user;
+
+  $type = is_string($node) ? $node : node_type_get_type($node);
+
+  // If this node is not defined by tripal_hq, tell drupal that we don't care
```
Comment (almasaeed2010) [inline tripal_hq.module]:
```suggestion
  // If this node is not defined by tripal_hq, tell drupal that we don't care.
```

File: tripal_hq.module [hunk 15]
```diff
@@ -192,3 +197,99 @@ function tripal_hq_entity_delete($entity, $type) {
...
+  if ($type !== 'tripal_hq_nd') {
+    return NODE_ACCESS_IGNORE;
+  }
+
+  // The user is an admin. They shall pass!
+  if (user_access('access tripal_hq admin')) {
+    return NODE_ACCESS_ALLOW;
+  }
+
+  // Get the user id for the submission associated with this node
```
Comment (almasaeed2010) [inline tripal_hq.module]:
```suggestion
  // Get the user id for the submission associated with this node.
```

File: tripal_hq.module [hunk 16]
```diff
@@ -192,3 +197,99 @@ function tripal_hq_entity_delete($entity, $type) {
...
+  $submission = db_select('tripal_hq_submission', 'THS')
+    ->fields('THS', ['uid'])
+    ->condition('nid', $node->nid)
+    ->execute()
+    ->fetchObject();
+
+  // We don't have a submission associated with this node ... DENY!
+  if(!$submission) {
```
Comment (almasaeed2010) [inline tripal_hq.module]:
```suggestion
  if (!$submission) {
```

File: tripal_hq.module [hunk 17]
```diff
@@ -192,3 +197,99 @@ function tripal_hq_entity_delete($entity, $type) {
...
+    return NODE_ACCESS_DENY;
+  }
+
+  // Allow the user to create comments if they created the submission
```
Comment (almasaeed2010) [inline tripal_hq.module]:
```suggestion
  // Allow the user to create comments if they created the submission.
```

File: tripal_hq.module [hunk 18]
```diff
@@ -192,3 +197,99 @@ function tripal_hq_entity_delete($entity, $type) {
...
+  if ($op === 'view' && $user->uid == $submission->uid) {
+    return NODE_ACCESS_ALLOW;
+  }
+
+  // Deny all other operations to none admins
```
Comment (almasaeed2010) [inline tripal_hq.module]:
```suggestion
  // Deny all other operations to none admins.
```

File: tripal_hq.install [hunk 19]
```diff
@@ -76,3 +85,18 @@ function tripal_hq_uninstall() {
     db_drop_table('tripal_hq_submission');
   }
 }
+
+/**
+ * Add nid column to tripal_hq_submission table.
+ */
+function tripal_hq_update_2100() {
```
Comment (bradfordcondon) [inline tripal_hq.install]: should probably be 3100?

**Comment by bradfordcondon** [comment]
looks good, thanks!!!

This should probably say add/view comments
screen shot 2019-01-17 at 10 41 03 am

**Pull request merged**